Add Python 3 support for validate-modules test.

frenzymadness opened the pull request:
Add Python 3 support for validate-modules test.

Fix #18367

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
test/sanity/validate-modules

##### ANSIBLE VERSION
```
ansible 2.3.0
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY
With this change, the validate-modules test works with Python 2 and Python 3 and produces the same output with both versions of Python. I've tested module validation on all modules in lib/ansible/modules and it works.

sivel commented on test/sanity/validate-modules/utils.py:
```diff
@@ -57,8 +56,6 @@ def __enter__(self):
         sys.stdout = self.stdout = StringIO()
         sys.stderr = self.stderr = StringIO()
-        setattr(sys.stdout, 'encoding', self.sys_stdout.encoding)
-        setattr(sys.stderr, 'encoding', self.sys_stderr.encoding)
```
These lines were added on purpose, and will cause failure in Ansible itself.  The line above about not using cStringIO due to needing this setattr also eludes to this.

I didn't explicitly outline it in the original commit where it was added, but not having `encoding` on `sys.stdout` and `sys.stderr` caused a traceback in ansible itself when calling `get_docstring` when calling `Display.display`.

I'm not sure if python3 will require the fileobj to have an `encoding` attribute, but if it does, we'll probably have to look at using StringIO for py2, and BytesIO for py3.

frenzymadness replied:
Sorry for this but I thought that this piece of code is used only for module validation and to be honest I still cannot find the connection between `CaptureStd()` and mentioned `get_docstring()`. I also used grep on the whole codebase and I cannot find another usage of this class than the one in validate-modules.

The reason for this change was that `StringIO` from `io` module does not support attributes setting and it is handling everything as unicode which I thought is right.

I'll change it to `io.BytesIO` where encoding attribute setting is possible.

sivel replied:
The connection, is that when `DOCUMENTATION` is invalid, `get_docstring` will call `display.error` and `display.display`.  I don't 100% remember the cause, but when writing to those fake `.stdout` and `.stderr`, an error would be raised when they did not contain a `.encoding` attribute.

frenzymadness replied:
This is the hardest part but I think I finally got it.

Because StringIO and BytesIO in Python š don't support encoding, I'll use BytesIO in combination with TextIOWrapper because TextIOWrapper supports encoding specification.

sivel commented on test/sanity/validate-modules/validate-modules:
```diff
@@ -41,6 +41,14 @@ from schema import doc_schema, option_schema, metadata_schema
+
+if PY3:
+    # Because there is no ast.TryExcept in Python 3 ast module
+    ast.TryExcept = ast.Try
```
I think I would prefer instead of faking this attribute, we create a var like:

```
if PY3:
    TRY_EXCEPT = ast.Try
else:
    TRY_EXCEPT = ast.TryExcept
```

And then use `TRY_EXCEPT` in the code.

frenzymadness replied:
Ok, I am gonna change it.

sivel commented on test/sanity/validate-modules/validate-modules:
```diff
@@ -41,6 +41,14 @@ from schema import doc_schema, option_schema, metadata_schema
+    ast.TryExcept = ast.Try
+    # REPLACER_WINDOWS from ansible.executor.module_common is byte
+    # string but we need unicode for Python 3
+    REPLACER_WINDOWS = REPLACER_WINDOWS.decode('utf-8')
```
If I understand why this is being done, and I think I do, the better approach would be instead to open the module file with `rb` instead of changing the type here.

```
diff --git a/test/sanity/validate-modules/validate-modules b/test/sanity/validate-modules/validate-modules
index 6b626f5..f4488d6 100755
--- a/test/sanity/validate-modules/validate-modules
+++ b/test/sanity/validate-modules/validate-modules
@@ -145,7 +145,7 @@ class ModuleValidator(Validator):

         self._python_module_override = False

-        with open(path) as f:
+        with open(path, 'rb') as f:
             self.text = f.read()
         self.length = len(self.text.splitlines())
         try:
```

frenzymadness replied:
You cannot simply change mode to `rb` because there are a lot of code parts where comparison with the unicode is used like `'sys.exit(' in self.text`, `('GNU General Public License' not in self.text and 'version 3' not in self.text)`, `'WANT_JSON' not in self.text` and more.

I think that replacing all these strings with bytes and load module file as bytes is a bad idea because module files are unicode and should be loaded as unicode.

Possibilities:
* Use my code and just convert REPLACER_WINDOWS to unicode
* Get rid of REPLACER_WINDOWS which is used only once a replace it directly with the unicode string `u'# POWERSHELL_COMMON'` which is IMHO worse and going against DRY principe.

Do you have a better idea?

sivel commented on test/sanity/validate-modules/schema.py:
```diff
@@ -17,29 +17,33 @@
+from ansible.compat.six import string_types
+# Because string_types from six is tuple mostly used in isinstance()
+# function and we need just class name for schema definition
+string_types = string_types[0]
```
I feel like my preference here would be to not overwrite the same variable, it can cause confusion.

Instead maybe don't do this at all, and instead use `*string_types` everywhere instead.

sivel replied:
That may have to be `*string_types` and `list(string_types)`.

Maybe create `string_types_list = list(string_types)` so you avoid recasting to a list multiple times.

frenzymadness replied:
Ok, I am gonna change it. I'll have to add `Any()` on lines where only `*string_types` will be used and change the order of arguments because the one with the star has to be the last one.

frenzymadness committed:
Add Python 3 support for validate-modules test.

Fix #18367

frenzymadness commented:
I've added the new version of commit with addressed review.

frenzymadness commented:
Hello. Is there any problem standing against merge?

frenzymadness committed:
Improve PEP8 compatibility

frenzymadness committed:
Fix Python 3 incompatibility

Is prohibited to mutate OrderedDict during iteration through it so
is better to add records with error or warning to empty dictionary
instead of delete records from copy of dictionary during iterating.

frenzymadness commented:
Hello.

I've resolved the conflict and rebased the branch.

I've also fixed one more Python 3 incompatibility which came with @sivel last change in https://github.com/ansible/ansible/commit/c160ede789b0c2f5361cddc07fbab2dccb9adf10 - `RuntimeError: OrderedDict mutated during iteration`. Fix is in separated commit.

Is possible to merge this PR before another conflict or incompatibility appears?

Thank you.

gundalow commented:
Hi, not see this PR before, will need a rebase as I've had to make a few changes to `schema.py`

frenzymadness committed:
Decode output of subprocess from bytes to unicode.

frenzymadness commented:
Hi @gundalow 

My branch is rebased and tested again. I fixed all issues after rebase and output of this test is the same for Python 2 and Python 3 (actually there are some differences but only reasonable like <str> vs <unicode> in traces etc.).

frenzymadness commented:
Are the failing tests problem? IMHO no, because the output of validate-modules seems to be valid. But I am not sure.

gundalow commented:
I think your validation of `description` has issues which is why it's throwing up errors for most (all?) modules

frenzymadness commented:
Thanks, I am gonna check it.

frenzymadness commented:
The bug is found and fixed now.

abadger commented on test/sanity/validate-modules/schema.py:
```diff
@@ -17,52 +17,62 @@
 #    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 from voluptuous import PREVENT_EXTRA, Any, Required, Schema
+from ansible.compat.six import string_types
```
ansible.compat.six has gone away in devel.  Now you can use ansible.module_utils.six everywhere.

abadger commented:
@frenzymadness Found one thing that needs updating for latest devel.

@sivel @gundalow  @mattclay Could I get a final review from any of you to merge this?

frenzymadness commented:
Hi @abadger 

I rebased the PR to current devel, switched from `compat.six` to `module_utils.six ` and tested it again in Python 2 and Python 3.

Could somebody please merge it before some other incompatibility or conflict appears? It is opened for more than two months. Thank you!

abadger commented:
Merged to devel.  Thanks @frenzymadness for porting that!  Thanks @sivel and @gundalow for reviewing.
